README.md
[![Build Status](https://travis-ci.org/Zaab1t/datatyping.svg?branch=master)](https://travis-ci.org/Zaab1t/datatyping)


# datatyping
datatyping lets you define a data structure of types and then verify that your
data is well-formed!

``` python
>>> import datatyping
>>> structure = {
...     'response_code': int,
...     'cars': [{'model': str, 'passengers': int}],
... }
>>> data = {
...     'response_code': 215,
...     'cars': [
...         {'model': 'Cadillac', 'passengers': 2},
...         {'model': 'Volvo', 'passengers': 4},
...     ]
... }
>>> datatyping.validate(structure, data)
```


## Install
Latest release:
`pip install datatyping`

Development version:
`pip install git+https://github.com/Zaab1t/datatyping`


## Benefits
- Documentation of incoming data in source code.
- Good for testing (especially if you offer something like a json api).
- Early failure in a specific spot if data is malformed or changed.
- Readable, explicit code base helps maintainability.


## Features
### Basics
``` python
>>> validate([int, str], [1, 'a'])
>>> validate([[int], [str]], [[1, 2, 3], ['a', 'b', 'c'])
```

``` python
>>> validate([dict], [{'can have': 1}, {'any keys': 2}])
>>> validate({'a': int, 'b': str}, {'a': 4, 'b': 'c'})
>>> validate({'a': int}, {'a': 2, 'b': 'oops'})
KeyError: {'b'}
>>> validate({'a': int}, {'a': 2, 'b': 'yay'}, strict=False)
```

### Custom types
``` python
>>> from datatyping import validate, customtype
>>> @customtype
>>> def positive_int(i):
...     if i < 1:
...         raise TypeError('%d is not positive' % i)

>>> validate([positive_int], [1, 2, 3, 4])
>>> validate([positive_int], [1, 2, 3, -4])
TypeError: -4 is not positive
```

### Structure generation
Since typing out the structure can be quite tedious, `datatyping` provides a tool that does most of the work for you:

``` python
>>> from datatyping.printer import pprint
>>> import requests
>>> r = requests.get('http://httpbin.org/anything')
>>> pprint(r.json())
{   
    'args': dict,
    'data': str,
    'files': dict,
    'form': dict,
    'headers': {   
        'Accept': str,
        'Accept-Encoding': str,
        'Connection': str,
        'Host': str,
        'User-Agent': str,
    },
    'json': NoneType,
    'method': str,
    'origin': str,
    'url': str,
}

```


## Testimonials
> does the data good
>> -- [theelous3](https://github.com/theelous3)


## Develop with me :)

``` bash
# fork it
$ git clone https://github.com/your_name/datatyping
$ cd datatyping
$ sudo python3 setup.py develop
# make your changes, commit and push
$ python3 -m pytest
# submit a pull request
```


## Notes
- Inspired by ["How Python Makes Working With Data More Difficult in the Long Run"](https://jeffknupp.com/blog/2016/11/13/how-python-makes-working-with-data-more-difficult-in-the-long-run/).
- Any and all contributions are welcome.
- Please open an issue if there's anything you can't make work (read: I messed up).
- Please let me know if there is an unsupported data structure, you'd like to see support for.
- If you enjoy and use the software, you can [say thanks](https://saythanks.io/to/Zaab1t) :)


CHANGELOG.md
# Changelog
## 0.6.0 (2017-11-09)
- Better exception messages with `reprlib.repr`
- External documentation (http://datatyping.rtfd.io)
- Checks that types are actually the same or a subclass instead of just of the
    same abstract type Sequence/Mapping


## 0.5.0 (2017-11-09)
- Replaced "contracts" with @customtype
- Works with all `collections.abc.Mapping` instead of just dicts.
- `validate` throws ValueError instead of undocumented AssertionError.
- New description: "Pythonic type checking"
- First PyPI release!


## 0.4.0 (2017-09-18)
- Added support for tuples
- Fixed bug with str (not considered sequence anymore)
- Reasonble versioning scheme :p


## 0.0.3 (2017-08-20)
- Added printer


## 0.0.2 (2017-07-27)
- Added contracts


## 0.0.1 (2017-07-26)
- Initial release


test_.py
import pytest
from datatyping import validate


def test_empty():
    assert validate({}, {})


def test_plain():
    assert validate({'a': int, 'b': str}, {'a': 1, 'b': 'c'})


def test_plain_typeerror():
    with pytest.raises(TypeError):
        validate({'a': int}, {'a': 3.4})


def test_plain_keyerror():
    with pytest.raises(KeyError):
        validate({'a': str, 'b': float}, {'a': 'abc'})


def test_plain_no_strict():
    assert validate({'a': str}, {'a': 'abc', 'b': 123}, strict=False)


def test_plain_no_strict_error():
    with pytest.raises(KeyError):
        validate({'a': str, 'b': float}, {'a': 'abc'}, strict=False)


def test_strict():
    with pytest.raises(KeyError):
        validate({'a': str}, {'a': 'abc', 'oh no too much data': 123})


def test_list():
    assert validate({'a': [int]}, {'a': [1, 2, 3]})


def test_list_typeerror():
    with pytest.raises(TypeError):
        validate({'a': [int]}, {'a': [1, 2, 3.6]})


def test_nested_list():
    assert validate({'a': [[int], [str]]},
            {'a': [[1,2,3,4], ['a', 'b', 'c', 'd']]})


def test_nested_dict():
    assert validate({'a': {'b': [[int], [str]]}},
            {'a': {'b': [[1,2,3,4], ['a','b','c']]}})


def test_mad_nesting():
    assert validate({'list': [int, str, [[int], [int]]]},
            {'list': [1, 'a', [[1,2,3], [4,5,6]]]})


def test_advanced():
    structure = {
        'total_count': int,
        'active': bool,
        'people': [
            {
                'id': int,
                'title': str,
                'salary': float,
            }
        ],
        'timestamps': [int],
    }
    data = {
        'total_count': 52,
        'active': False,
        'people': [
            {'id': 23974291847, 'title': 'big guy', 'salary': 256.12},
            {'id': 1932, 'title': 'python developer', 'salary': 4.5},
            {'id': 9828974, 'title': 'president', 'salary': 122352.23},
        ],
        'timestamps': [12351346, 12345134, 7684545, 1267457, 0],
    }
    assert validate(structure, data)


tests/test_str.py
import pytest
from datatyping import validate


def test_simple():
    assert validate([str], ['ab', 'cd']) is None


def test_simple_error():
    with pytest.raises(TypeError):
        validate([str], 'asd')


def test_nested():
    assert validate([[str]], [['ab', 'cd'], ['ef', 'g']]) is None


test_dict_struct.py
import pytest
from datatyping import validate


def test_empty():
    assert validate({}, {})


def test_plain():
    assert validate({'a': int, 'b': str}, {'a': 1, 'b': 'c'})


def test_plain_typeerror():
    with pytest.raises(TypeError):
        validate({'a': int}, {'a': 3.4})


def test_plain_keyerror():
    with pytest.raises(KeyError):
        validate({'a': str, 'b': float}, {'a': 'abc'})


def test_plain_no_strict():
    assert validate({'a': str}, {'a': 'abc', 'b': 123}, strict=False)


def test_plain_no_strict_error():
    with pytest.raises(KeyError):
        validate({'a': str, 'b': float}, {'a': 'abc'}, strict=False)


def test_strict():
    with pytest.raises(KeyError):
        validate({'a': str}, {'a': 'abc', 'oh no too much data': 123})


def test_list():
    assert validate({'a': [int]}, {'a': [1, 2, 3]})


def test_list_typeerror():
    with pytest.raises(TypeError):
        validate({'a': [int]}, {'a': [1, 2, 3.6]})


def test_nested_list():
    assert validate({'a': [[int], [str]]},
        {'a': [[1,2,3,4], ['a', 'b', 'c', 'd']]})


def test_nested_dict():
    assert validate({'a': {'b': [[int], [str]]}},
        {'a': {'b': [[1,2,3,4], ['a','b','c']]}})


def test_mad_nesting():
    assert validate({'list': [int, str, [[int], [int]]]},
        {'list': [1, 'a', [[1,2,3], [4,5,6]]]})


def test_advanced():
    structure = {
        'total_count': int,
        'active': bool,
        'people': [
            {
                'id': int,
                'title': str,
                'salary': float,
            }
        ],
        'timestamps': [int],
    }
    data = {
        'total_count': 52,
        'active': False,
        'people': [
            {'id': 23974291847, 'title': 'big guy', 'salary': 256.12},
            {'id': 1932, 'title': 'python developer', 'salary': 4.5},
            {'id': 9828974, 'title': 'president', 'salary': 122352.23},
        ],
        'timestamps': [12351346, 12345134, 7684545, 1267457, 0],
    }
    assert validate(structure, data)


datatyping/printer.py
"""Like pprint, but with types except for dictionary keys."""


__all__ = ["pprint", "pformat"]
__author__ = "Carl Bordum Hansen"
__license__ = "MIT"


import pprint as _pprint
import contextlib


def _new_format_dict_items(self, items, stream, indent, allowance, context, level):
    write = stream.write
    indent += self._indent_per_level
    last_index = len(items) - 1
    write("\n")
    write(" " * indent)
    for i, (key, ent) in enumerate(items):
        last = i == last_index
        rep = repr(key)
        write(rep)
        write(": ")
        self._format(ent, stream, indent, allowance if last else 1, context, level)
        write(",\n")
        if last:
            write((" " * indent)[: -self._indent_per_level])
        else:
            write(" " * indent)


def _new_safe_repr(object, context, maxlevels, level):
    """Return object type name except for dict keys.
    
    Like `pprint._safe_repr`, but returns type name of object instead
    of object repr except for dictionary keys. Also formats lists and
    tuples nicely.

    Used to overwrite `pprint._safe_repr` with the context manager
    `change_pprint_repr`.
    """
    typerepr = lambda object: type(object).__name__
    type_ = type(object)
    if type_ in _pprint._builtin_scalars:
        return typerepr(object), True, False

    r = getattr(type_, "__repr__", None)
    if issubclass(type_, dict) and r is dict.__repr__:
        if not object:
            return "dict", True, False
        context[id(object)] = 1
        readable = True
        recursive = False
        level += 1
        pairs = []
        for k, v in object.items():
            vrepr, vreadable, recur = _new_safe_repr(v, context, maxlevels, level)
            pairs.append("%s: %s" % (repr(k), vrepr))
            readable = readable and vreadable
            if recur:
                recursive = True
        del context[id(object)]
        return "{%s}" % ", ".join(pairs), readable, recursive
    if issubclass(type_, (list, tuple)):
        if issubclass(type_, list):
            if not object:
                return "list", True, False
            format = "[%s]"
        else:  # its a tuple
            if not object:
                return "tuple", True, False
            format = "(%s)" if len(object) != 1 else "(%s,)"
        context[id(object)] = 1
        readable = True
        recursive = False
        items = []
        level += 1
        for item in object:
            irepr, ireadable, irecur = _new_safe_repr(item, context, maxlevels, level)
            items.append(irepr)
            if not ireadable:
                readable = False
            if irecur:
                recursion = True
        del context[id(object)]
        if len(set(items)) == 1:
            items = [items[0]]
        return format % ", ".join(items), readable, recursive
    return typerepr(object), True, False


@contextlib.contextmanager
def change_pprint_repr():
    old_safe_repr = _pprint._safe_repr
    _pprint._safe_repr = _new_safe_repr
    try:
        old_format_dict_items = _pprint.PrettyPrinter._format_dict_items
    except AttributeError:
        raise RuntimeError("Only Python3.5+ supported (datatyping.printer)")
    _pprint.PrettyPrinter._format_dict_items = _new_format_dict_items
    yield
    _pprint._safe_repr = old_safe_repr
    _pprint.PrettyPrinter._format_dict_items = old_format_dict_items


def pprint(object, stream=None, indent=4, width=80, depth=None, compact=False):
    """Pretty-prints the data structure."""
    with change_pprint_repr():
        _pprint.pprint(
            object,
            stream=stream,
            indent=indent,
            width=width,
            depth=depth,
            compact=compact,
        )


def pformat(object, indent=4, width=80, depth=None, compact=False):
    """Return the pretty printed data structure of *object*."""
    with change_pprint_repr():
        return _pprint.pformat(
            object, indent=indent, width=width, depth=depth, compact=compact
        )


datatyping/datatyping.py
__all__ = ["validate", "customtype"]
__author__ = "Carl Bordum Hansen"
__license__ = "MIT"


import collections.abc
import reprlib


# Open questions:
#     how should sets be handled?
#     should we handle collections.abc.MappingView?


def _make_typeerror(structure, data):
    error_msg = "%s is of type %s, expected type %s" % (
        reprlib.repr(data),
        data.__class__.__name__,
        structure.__class__.__name__,
    )
    return TypeError(error_msg)


def customtype(check_function):
    """Decorate a function, so it can be used for type checking.

    Example
    -------
        >>> @customtype
        ... def two_item_list(lst):
        ...     if len(lst) != 2:
        ...         raise TypeError('length %d!!!' % len(lst))
        ...
        >>> validate([two_item_list], [[1, 2], [3, 4]])
        >>> validate([two_item_list], [[1, 2], [3, 4, 5]])
        TypeError: length 3!!!

    Note
    ----
        Sets the `check_function.__datatyping_validate` attribute.

    Parameters
    ----------
    check_function : function
        Function that should be used to type check.

    """
    check_function.__datatyping_validate = True
    return check_function


def validate(structure, data, *, strict=True):
    """Verify that values in a dataset are of correct types.

    Example
    -------
        >>> validate([str], ['a', 'b', 'c'])
        >>> validate(
        ...     {'id': int, 'lucky_numbers': [int]},
        ...     {'id': 700, 'lucky_numbers': [1, 3, 7, 13]}
        ... )
        >>> validate([int], [1, 2, 3, 4.5])
        TypeError: 4.5 is of type float, expected type int.

    Parameters
    ----------
    structure : type or collection of types
        The data structure that `data` should follow.
    data : anything
        The data you want type checked.
    strict : bool
        Dicts in `data` must have the **exact** keys specified in
        `structure`. No more.

    Raises
    ------
    TypeError
        If an elements in `data` has wrong type.
    ValueError
        If the length of `structure` doesn't make sense for validating
        `data`.
    KeyError
        If a dict in `data` misses a key or `strict` is True and a dict
        has keys not in `structure`.

    """
    if hasattr(structure, "__datatyping_validate"):
        structure(data)
    elif isinstance(structure, collections.abc.Sequence) and not isinstance(data, str):
        if not isinstance(data, type(structure)):
            raise _make_typeerror(structure, data)
        if len(structure) == 1:
            for item in data:
                validate(structure[0], item, strict=strict)
            return  # success
        if len(structure) != len(data):
            error_msg = ("%s has the wrong length. Expected %d, got %d.") % (
                reprlib.repr(data),
                len(structure),
                len(data),
            )
            raise ValueError(error_msg)
        for type_, item in zip(structure, data):
            validate(type_, item, strict=strict)
    elif isinstance(structure, collections.abc.Mapping):
        if not isinstance(data, type(structure)):
            raise _make_typeerror(structure, data)
        if strict and len(structure) != len(data):
            raise KeyError(reprlib.repr(set(structure.keys()) ^ set(data.keys())))
        for key, type_ in structure.items():
            item = data[key]  # or KeyError :)
            validate(type_, item, strict=strict)
    elif not isinstance(data, structure):  # structure is a type here
        error_msg = "%s is of type %s, expected type %s" % (
            reprlib.repr(data),
            data.__class__.__name__,
            structure.__name__,
        )
        raise TypeError(error_msg)
